internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,668 @@
+					producer.iter.Stop()
+					producer.iter = nil
+					continue
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			tmpKey := t.GetObject()
+			for tmpKey < maxKey {
+				t, err := producer.iter.Next(ctx)
+				if err != nil {
+					if storage.IterIsDoneOrCancelled(err) {
+						producer.iter.Stop()
+						producer.iter = nil
+						break
+					}
+					concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+					return
+				}
+				tmpKey = t.GetObject()
+			}
+		}
+	}
+}
+
+func fastPathDifference(ctx context.Context, iterProducers []*iteratorProducer, iterChan chan *iteratorMsg) {
+	batch := make([]*openfgav1.TupleKey, 0)
+
+	defer func() {
+		// flush
+		if len(batch) > 0 {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: storage.NewStaticTupleKeyIterator(batch)}, iterChan)
+		}
+		close(iterChan)
+		cleanupIteratorProducers(iterProducers)
+	}()
+
+	for len(iterProducers) == 2 {
+		if ctx.Err() != nil {
+			return
+		}
+		var err error
+		iterProducers, err = pollIteratorProducers(ctx, iterProducers)
+		if err != nil {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+			return
+		}
+		if len(iterProducers) != 2 {
+			// short circuit
+			break
+		}
+
+		allIters := true
+		base := ""
+		diff := ""
+		for idx, producer := range iterProducers {
+			v, err := producer.iter.Head(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					producer.iter.Stop()
+					producer.iter = nil
+					allIters = false
+					// we need to ensure we have all iterators at all times
+					break
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			if idx == 0 {
+				base = v.GetObject()
+			}
+			if idx == 1 {
+				diff = v.GetObject()
+			}
+		}
+
+		if !allIters {
+			// we need to ensure we have all iterators at all times
+			continue
+		}
+
+		// move both heads
+		if base == diff {
+			for _, iter := range iterProducers {
+				_, err := iter.iter.Next(ctx)
+				if err != nil {
+					if storage.IterIsDoneOrCancelled(err) {
+						iter.iter.Stop()
+						iter.iter = nil
+						break
+					}
+					concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+					return
+				}
+			}
+			continue
+		}
+
+		if diff > base {
+			t, err := iterProducers[0].iter.Next(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					iterProducers[0].iter.Stop()
+					iterProducers[0].iter = nil
+					break
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			batch = append(batch, t)
+			if len(batch) > IteratorBatchThreshold {
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: storage.NewStaticTupleKeyIterator(batch)}, iterChan)
+				batch = make([]*openfgav1.TupleKey, 0)
+			}
+			continue
+		}
+
+		// base > diff, then move the diff to catch up with base
```

Comment:
```suggestion
		// diff < base, then move the diff to catch up with base
```
using the same term for comment and code makes it easier to read